### DISCUSSION TITLE
Solve PDE in Implicit Method

### DISCUSSION THREAD

**Discussion by beantsja** [zwicker-group/py-pde#508] [Q&A]
Hi, I am trying to solve simple 1D Heat Equation, but I want the length of the rod to be around 1 micrometer and the total spatial grid points to be around 25. Under this condition, it is forces dt to be extremely small under explicit method, and diffusionPDE seems to use that method. 

I noticed there is Implicit Solver in the modules. Based on that, how would I solve PDE using implicit method?

**Comment by david-zwicker** [comment]
You can simply specify `eq.solve(..., solver="implicit")` to pick an implicit solver; see [documentation](https://py-pde.readthedocs.io/en/latest/packages/pde.pdes.base.html#pde.pdes.base.PDEBase.solve). If you want to have more control over the solver, you can also follow [this example](https://py-pde.readthedocs.io/en/latest/examples_gallery/advanced_pdes/solver_comparison.html).

However, it might also be a good idea to increase the spatial discretization to accelerate the procedure.

**Comment by DelightE99** [comment]
I am trying to follow the example to have more control over the implicit solver, but the link seems not to be available when I click it. Kindly see below

![image](https://github.com/user-attachments/assets/72af0e99-9ef8-49f5-8cfb-0e3087991928)

**Reply by david-zwicker** [reply]
The example has been moved to https://py-pde.readthedocs.io/en/latest/examples_gallery/advanced_pdes/solver_comparison.html.

**Reply by DelightE99** [reply]
Thank you, but the examples (see below) show mostly explicit solvers and not implicit, except I'm missing something?

import pde

# initialize the grid, an initial condition, and the PDE
grid = pde.UnitGrid([32, 32])
field = pde.ScalarField.random_uniform(grid, -1, 1)
eq = pde.DiffusionPDE()

# try the explicit solver
solver1 = pde.ExplicitSolver(eq)
controller1 = pde.Controller(solver1, t_range=1, tracker=None)
sol1 = controller1.run(field, dt=1e-3)
sol1.label = "explicit solver"
print("Diagnostic information from first run:")
print([controller1.diagnostics](https://docs.python.org/3/library/stdtypes.html#dict))
print()

# try an explicit solver with adaptive time steps
solver2 = pde.ExplicitSolver(eq, scheme="runge-kutta", adaptive=True)
controller2 = pde.Controller(solver2, t_range=1, tracker=None)
sol2 = controller2.run(field, dt=1e-3)
sol2.label = "explicit, adaptive solver"
print("Diagnostic information from second run:")
print([controller2.diagnostics](https://docs.python.org/3/library/stdtypes.html#dict))
print()

# try the standard scipy solver
solver3 = pde.ScipySolver(eq)
controller3 = pde.Controller(solver3, t_range=1, tracker=None)
sol3 = controller3.run(field)
sol3.label = "scipy solver"
print("Diagnostic information from third run:")
print([controller3.diagnostics](https://docs.python.org/3/library/stdtypes.html#dict))
print()

# plot both fields and give the deviation as the title
[deviation12](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.float64) = ((sol1 - sol2) ** 2).average
[deviation13](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.float64) = ((sol1 - sol3) ** 2).average
[title](https://docs.python.org/3/library/stdtypes.html#str) = f"Deviation: {[deviation12](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.float64):.2g}, {[deviation13](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.float64):.2g}"
pde.FieldCollection([sol1, sol2, sol3]).plot([title](https://docs.python.org/3/library/stdtypes.html#str)=[title](https://docs.python.org/3/library/stdtypes.html#str))

**Reply by DelightE99** [reply]
I tried following the example to edit the implicit solver viz:

    #Defining PDE with boundary conditions
    eq = PDE({
        "T_f": f" ({Df}  * laplace(T_f))  +  ({-Cf}  * d_dx(T_f)) + ({Alpha_f}  * (T_s - T_f)) + ({Beta_f}  * ({Tamb} - T_f))",
        "T_s": f" ( {Ds}  * laplace(T_s)) + ( {Alpha_s}  * (T_f - T_s))"
    }, bc_ops = bc_total)

    # Simulation
    # Solve the coupled PDEs with initial conditions
    #result = eq.solve(Initial_conditions, t_range=time_simulated, dt=dt, solver="explicit")

    # Solve the PDE using the implicit solver with specified parameters

    #Specify Solver parameters
    solver2 = ImplicitSolver(eq, maxiter=10000, maxerror=0.1, backend='auto')
    controller2 = Controller(solver2, t_range=time_simulated, tracker=None)
    sol2 = controller2.run(Initial_conditions, dt=time_simulated/10)
    sol2.label = "implicit modified"

    # Solve the coupled PDEs with initial conditions
    result = eq.solve(
        Initial_conditions,
        t_range=time_simulated,
        dt=time_simulated/10,
        solver="implicit modified",       # Use the implicit solver
        #max_iterations=1000,     # Maximum number of iterations
        #tol=1e-2,                # Convergence tolerance
        #backend="numpy",         # Computational backend
        #initial_guess=Initial_conditions  # Initial guess for the solution
    )
    print("Implicit solver finished")

However, I keep getting the error "Implicit solver not converged" no matter how I improve the iteration parameters. I am using a spatial discretization of 200 over a 42m length and I do not know whether this could be the problem. Also, how do I specify the initial guess that the solver starts with, etc.

**Reply by david-zwicker** [reply]
Unfortunately I don't have the bandwidth to investigate convergence problems of specific equations.

**Reply by DelightE99** [reply]
Yeah sure, Thank you so much for the directives. The main issue really was to confirm the structure of the code implemented for the control of the implicit solver. The rest would follow eventually.

**Reply by david-zwicker** [reply]
You can certainly implement your own implicit solver if you want to have more control. A good starting point might be the implementation of the [`CrankNicolsonSolver`](https://github.com/zwicker-group/py-pde/blob/master/pde/solvers/crank_nicolson.py).
